### PR TITLE
feat(java): Migrate `pom.xml` to JDK8.

### PR DESCRIPTION
**Pull request by gounthar**
# Discussion: Brutal POM Modification for Legacy Plugins

## Purpose of this PR

This pull request is not intended for merging. It serves as a discussion base and a potential source for cherry-picking ideas. The PR contains merge conflicts due to recent changes (e.g., Guice) that have not been integrated.

## Goal of the Code Modification

The proposed code aims to forcefully modify the `pom.xml` file when:
1. We detect a JDK version older than 8
2. The `--ignore-jdk7-error` flag is used on the command line

## Modifications Made

When conditions are met, the code:
1. Changes the POM parent
2. Updates the Jenkins version
3. Removes the `jdk.version` property
4. Adds the BOM (Bill of Materials)

## Limitations

This approach is not a universal solution and may not work for every plugin. However, it successfully upgraded the [outdated Vagrant plugin](#215) and could serve as a foundation for future improvements.

## Testing Performed

The modification was tested using the following command:
`java -jar plugin-modernizer-cli/target/jenkins-plugin-modernizer-999999-SNAPSHOT.jar --plugins vagrant --recipes MinimalBuildJava8 --ignore-jdk7-error --debug --dry-run`
Additionally, the changes have been verified to work on GitPod after setting the appropriate environment variables. You can test it [here](https://gitpod.io/?autostart=true#https://github.com/gounthar/plugin-modernizer-tool/tree/add-ignore-jdk7-error).

## Next Steps

1. Discuss the approach and its potential implications
2. Identify other plugins that could benefit from this method
3. Document known limitations and edge cases
4. Consider integrating successful aspects into the main modernization workflow

Your feedback and suggestions for improving this approach are welcome.

### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- ~~Link to relevant pull requests, esp. upstream and downstream changes~~
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/config/Config.java [hunk 1]
```diff
@@ -252,6 +260,11 @@ public Builder withExportDatatables(boolean exportDatatables) {
             return this;
         }
 
+        public Builder withIgnoreJdk7Error(boolean ignoreJdk7Error) {
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/config/Config.java:263]: I think this flag becomes obsolete if by default we try to fix precondition error

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/impl/PluginModernizer.java [hunk 2]
```diff
@@ -152,6 +155,23 @@ private void process(Plugin plugin) {
                 LOG.debug("Metadata already computed for plugin {}. Using cached metadata.", plugin.getName());
             }
 
+            // Check for ignoreJdk7Error option and use PomModifier if set
+            if (config.isIgnoreJdk7Error() && plugin.hasPreconditionErrors()) {
+                LOG.info("Ignoring JDK7 error for plugin {} and modifying pom.xml", plugin.getName());
+                PomModifier pomModifier = new PomModifier(
+                        plugin.getLocalRepository().resolve("pom.xml").toString());
```
Comment (jonesbusy) [inline plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/impl/PluginModernizer.java:162]: some pseudo code

```
plugin.getPreconditionErrors().each { precondition ->
    Document document = precondition.remediate()
    write(document,  plugin.getLocalRepository().resolve("pom.xml"))
}
```

**Comment by gounthar** [comment]
I was sure you would come up with something way better. 🤗
Thanks 👍

**Comment by jonesbusy** [comment]
I've added https://github.com/jenkinsci/plugin-modernizer-tool/pull/310

Perhaps can you start to move some of the logic of this PR on the remediation method when merged ?

File: plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/impl/PluginModernizer.java (hunk 2 above)
Comment (jonesbusy) [reply plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/impl/PluginModernizer.java:162]: On https://github.com/jenkinsci/plugin-modernizer-tool/pull/310

**Pull request closed without merging**